File: filefn/server/src/events.ts
import { EventEmitter } from 'events';
import type { ObservationEvent, SuperfunctionObservability } from '@superfunctions/observability';

export interface FileFnEvent {
  type: string;
  timestamp: string;
  requestId?: string;
}

const SECRET_EVENT_PATTERNS = [
  /https?:\/\/[^\s"']+\?[^\s"']*(?:X-Amz-Signature|Signature|sig|token|key)=[^\s"'&]+/gi,
  /[?&](?:X-Amz-Signature|Signature|sig|token|key)=[^\s"'&]+/gi,
  /Bearer\s+[A-Za-z0-9\-_]+\.[A-Za-z0-9\-_]+\.[A-Za-z0-9\-_]+/gi,
  /upls_live_[A-Za-z0-9\-_]+/g,
];

function shouldRedactEventKey(key: string): boolean {
  const lowerKey = key.toLowerCase();
  return (
    lowerKey.includes('token') ||
    lowerKey.includes('secret') ||
    lowerKey.includes('password') ||
    lowerKey.includes('signature') ||
    lowerKey.includes('signedurl') ||
    lowerKey.includes('signed_url') ||
    lowerKey.includes('authorization')
  );
}

function sanitizeEventValue(key: string, value: unknown): unknown {
  if (value === null || value === undefined) return value;
  if (shouldRedactEventKey(key)) return '[REDACTED]';

  if (typeof value === 'string') {
    let sanitized = value;
    for (const pattern of SECRET_EVENT_PATTERNS) {
      sanitized = sanitized.replace(pattern, '[REDACTED]');
    }
    return sanitized;
  }

  if (Array.isArray(value)) {
    return value.map((item) => sanitizeEventValue(key, item));
  }

  if (typeof value === 'object') {
    return sanitizeEventData(value as Record<string, unknown>);
  }

  return value;
}

function sanitizeEventData<T extends Record<string, unknown>>(data: T): T {
  const sanitized: Record<string, unknown> = {};
  for (const [key, value] of Object.entries(data)) {
    sanitized[key] = sanitizeEventValue(key, value);
  }
  return sanitized as T;
}

export interface UploadStartedEvent extends FileFnEvent {
  type: 'upload.started';
  uploadSessionId: string;
  fileName: string;
  size: number;
  mimeType: string;
  policy: string;
  principalId?: string;
  tenantId?: string;
}

export interface PartRecordedEvent extends FileFnEvent {
  type: 'part.recorded';
  uploadSessionId: string;
  partNumber: number;
  size: number;
}

export interface FileUploadedEvent extends FileFnEvent {
  type: 'file:uploaded';
  fileId: string;
  versionId: string;
  fileName: string;
  size: number;
  mimeType: string;
  ownerId: string;
  tenantId?: string;
}

export interface FileDeletedEvent extends FileFnEvent {
  type: 'file:deleted';
  fileId: string;
  ownerId: string;
  tenantId?: string;
}

export interface ProcessingStartedEvent extends FileFnEvent {
  type: 'processing.started';
  fileId: string;
  versionId: string;
}

export interface ProcessingCompletedEvent extends FileFnEvent {
  type: 'processing.completed';
  fileId: string;
  versionId: string;
  artifactsCreated: number;
}

export interface ProcessingFailedEvent extends FileFnEvent {
  type: 'processing.failed';
  fileId: string;
  versionId: string;
  error?: string;
}

export type FileFnEventTypes = {
  'upload.started': UploadStartedEvent;
  'part.recorded': PartRecordedEvent;
  'file:uploaded': FileUploadedEvent;
  'file:deleted': FileDeletedEvent;
  'processing.started': ProcessingStartedEvent;
  'processing.completed': ProcessingCompletedEvent;
  'processing.failed': ProcessingFailedEvent;
};

export type FileFnEventType = keyof FileFnEventTypes;
export type FileFnObservationMetadata<TEvent extends FileFnEvent> =
  Record<string, unknown> & Omit<TEvent, 'type' | 'requestId'>;

export type FileFnObservationEventMap = {
  [TType in FileFnEventType]: ObservationEvent<
    'filefn',
    TType,
    FileFnObservationMetadata<FileFnEventTypes[TType]>
  >;
};

export type FileFnObservationEvent = FileFnObservationEventMap[FileFnEventType];

export class FileFnEventEmitter extends EventEmitter {
  constructor(private readonly observability?: SuperfunctionObservability<FileFnObservationEvent>) {
    super();
  }

  emit<K extends keyof FileFnEventTypes>(event: K, payload: FileFnEventTypes[K]): boolean {
    void this.observability?.events.emit(fileObservationEvent(event, payload));
    return super.emit(event, payload);
  }

  on<K extends keyof FileFnEventTypes>(event: K, listener: (payload: FileFnEventTypes[K]) => void): this {
    return super.on(event, listener);
  }

  once<K extends keyof FileFnEventTypes>(event: K, listener: (payload: FileFnEventTypes[K]) => void): this {
    return super.once(event, listener);
  }
}

export function createUploadStartedEvent(
  data: Omit<UploadStartedEvent, 'type' | 'timestamp' | 'requestId'>,
  requestId?: string
): UploadStartedEvent {
  const sanitizedData = sanitizeEventData(data);
  return {
    type: 'upload.started',
    timestamp: new Date().toISOString(),
    requestId,
    ...sanitizedData,
  };
}

export function createPartRecordedEvent(
  data: Omit<PartRecordedEvent, 'type' | 'timestamp' | 'requestId'>,
  requestId?: string
): PartRecordedEvent {
  const sanitizedData = sanitizeEventData(data);
  return {
    type: 'part.recorded',
    timestamp: new Date().toISOString(),
    requestId,
    ...sanitizedData,
  };
}

export function createFileUploadedEvent(
  data: Omit<FileUploadedEvent, 'type' | 'timestamp' | 'requestId'>,
  requestId?: string
): FileUploadedEvent {
  const sanitizedData = sanitizeEventData(data);
  return {
    type: 'file:uploaded',
    timestamp: new Date().toISOString(),
    requestId,
    ...sanitizedData,
  };
}

export function createFileDeletedEvent(
  data: Omit<FileDeletedEvent, 'type' | 'timestamp' | 'requestId'>,
  requestId?: string
): FileDeletedEvent {
  const sanitizedData = sanitizeEventData(data);
  return {
    type: 'file:deleted',
    timestamp: new Date().toISOString(),
    requestId,
    ...sanitizedData,
  };
}

export function createEventEmitter(
  observability?: SuperfunctionObservability<FileFnObservationEvent>
): FileFnEventEmitter {
  return new FileFnEventEmitter(observability);
}

function fileObservationEvent<K extends FileFnEventType>(
  event: K,
  payload: FileFnEventTypes[K]
): FileFnObservationEvent {
  const observation: ObservationEvent<
    'filefn',
    K,
    FileFnObservationMetadata<FileFnEventTypes[K]>
  > = {
    domain: 'filefn',
    type: event,
    severity: event.includes('failed') ? 'error' : 'info',
    requestId: payload.requestId,
    metadata: fileEventMetadata(payload),
  };
  return observation as FileFnObservationEvent;
}

function fileEventMetadata<TEvent extends FileFnEvent>(
  event: TEvent
): FileFnObservationMetadata<TEvent> {
  const { type, timestamp, requestId, ...metadata } = event as FileFnEvent & Record<string, unknown>;
  return sanitizeEventData({
    timestamp,
    ...metadata
  }) as FileFnObservationMetadata<TEvent>;
}

File: filefn/python/filefn/server/events.py
from datetime import datetime, timezone
from typing import Any, Callable, Dict, List, Optional

from pydantic import BaseModel

SAFE_IDENTIFIER_KEYS = {
    "fileid",
    "versionid",
    "artifactid",
    "policyid",
    "requestid",
    "uploadsessionid",
}
SENSITIVE_KEY_MARKERS = (
    "token",
    "secret",
    "password",
    "signature",
    "authorization",
    "credential",
    "cookie",
    "signedurl",
    "signed_url",
)

class FileFnEvent(BaseModel):
    type: str
    timestamp: str
    requestId: Optional[str] = None

class UploadStartedEvent(FileFnEvent):
    type: str = 'upload.started'
    uploadSessionId: str
    fileName: str
    size: int
    mimeType: str
    policy: str
    principalId: Optional[str] = None
    tenantId: Optional[str] = None

class PartRecordedEvent(FileFnEvent):
    type: str = 'part.recorded'
    uploadSessionId: str
    partNumber: int
    size: int

class FileUploadedEvent(FileFnEvent):
    type: str = 'file:uploaded'
    fileId: str
    versionId: str
    fileName: str
    size: int
    mimeType: str
    ownerId: str
    tenantId: Optional[str] = None

class FileDeletedEvent(FileFnEvent):
    type: str = 'file:deleted'
    fileId: str
    ownerId: str
    tenantId: Optional[str] = None

class ProcessingStartedEvent(FileFnEvent):
    type: str = 'processing.started'
    fileId: str
    versionId: str

class ProcessingCompletedEvent(FileFnEvent):
    type: str = 'processing.completed'
    fileId: str
    versionId: str
    artifactsCreated: int

class ProcessingFailedEvent(FileFnEvent):
    type: str = 'processing.failed'
    fileId: str
    versionId: str
    error: Optional[str] = None

def get_current_timestamp() -> str:
    return datetime.now(timezone.utc).isoformat()

def create_upload_started_event(data: Dict[str, Any], request_id: Optional[str] = None) -> UploadStartedEvent:
    return UploadStartedEvent(
        timestamp=get_current_timestamp(),
        requestId=request_id,
        **data
    )

def create_part_recorded_event(data: Dict[str, Any], request_id: Optional[str] = None) -> PartRecordedEvent:
    return PartRecordedEvent(
        timestamp=get_current_timestamp(),
        requestId=request_id,
        **data
    )

def create_file_uploaded_event(data: Dict[str, Any], request_id: Optional[str] = None) -> FileUploadedEvent:
    return FileUploadedEvent(
        timestamp=get_current_timestamp(),
        requestId=request_id,
        **data
    )

def create_file_deleted_event(data: Dict[str, Any], request_id: Optional[str] = None) -> FileDeletedEvent:
    return FileDeletedEvent(
        timestamp=get_current_timestamp(),
        requestId=request_id,
        **data
    )

class FileFnEventEmitter:
    def __init__(self) -> None:
        self._listeners: Dict[str, List[Callable[[Any], None]]] = {}

    def on(self, event_type: str, listener: Callable[[Any], None]) -> 'FileFnEventEmitter':
        if event_type not in self._listeners:
            self._listeners[event_type] = []
        self._listeners[event_type].append(listener)
        return self

    def emit(self, event_type: str, payload: Any) -> bool:
        if event_type not in self._listeners:
            return False
        for listener in self._listeners[event_type]:
            try:
                listener(payload)
            except Exception:
                # Log error or suppress? Usually suppress in EventEmitter
                pass
        return True

def create_event_emitter() -> FileFnEventEmitter:
    return FileFnEventEmitter()


def _should_redact_key(key: str) -> bool:
    lower_key = key.lower()
    if lower_key in SAFE_IDENTIFIER_KEYS:
        return False
    return any(marker in lower_key for marker in SENSITIVE_KEY_MARKERS)


def _should_redact_string(key: str, value: str) -> bool:
    lower_key = key.lower()
    lower_value = value.lower()
    if lower_key in SAFE_IDENTIFIER_KEYS:
        return False
    if value.startswith("http://") or value.startswith("https://"):
        if "?" in value or "signature=" in lower_value or "sig=" in lower_value or "token=" in lower_value:
            return True
    if lower_value.startswith("bearer "):
        return True
    if "x-amz-signature" in lower_value:
        return True
    return False


def redact_secrets(value: Any, key: str = "") -> Any:
    if value is None:
        return None

    if isinstance(value, dict):
        result: Dict[str, Any] = {}
        for child_key, child_value in value.items():
            if _should_redact_key(child_key):
                result[child_key] = "[REDACTED]"
            else:
                result[child_key] = redact_secrets(child_value, child_key)
        return result

    if isinstance(value, list):
        return [redact_secrets(item, key) for item in value]

    if isinstance(value, tuple):
        return tuple(redact_secrets(item, key) for item in value)

    if isinstance(value, str) and _should_redact_string(key, value):
        return "[REDACTED]"

    return value
